multiapps-controller-persistence/src/main/java/org/cloudfoundry/multiapps/controller/persistence/services/DatabaseFileService.java

```diff
@@ -65,17 +65,17 @@ public int deleteFilesEntriesWithoutContent() throws FileStorageException {
     }
 
     @Override
-    protected void storeFile(FileEntry fileEntry, FileInfo fileinfo) throws FileStorageException {
-        try (InputStream fileStream = fileinfo.getInputStream()) {
-            storeFileWithContent(fileEntry, fileStream);
-        } catch (IOException e) {
-            logger.debug(e.getMessage(), e);
-        }
-    }
-
-    private boolean storeFileWithContent(FileEntry fileEntry, InputStream fileStream) throws FileStorageException {
+    protected FileEntry storeFile(FileEntry fileEntry, InputStream fileStream) throws FileStorageException {
         try {
-            return getSqlQueryExecutor().execute(getSqlFileQueryProvider().getStoreFileQuery(fileEntry, fileStream));
+            if (fileEntry.getDigest() != null) {
+                getSqlQueryExecutor().execute(getSqlFileQueryProvider().getStoreFileQuery(fileEntry, fileStream));
+                return fileEntry;
+            }
+            String digest = getSqlQueryExecutor().execute(getSqlFileQueryProvider().getStoreFileWithoutDigestQuery(fileEntry, fileStream));
+            getSqlQueryExecutor().execute(getSqlFileQueryProvider().getUpdateFileDigestQuery(fileEntry, digest));
```

Comment:
Is this in the same transaction and if not, can this be a problem? For example, having two people upload the same file in parallel. The first stores the content in the database successfully, but before updating the digest, the second person tries to retrieve the list of files to check whether he can skip the upload. He sees the file entry, tries to use the digest and fails with a nil pointer dereference, because it is nil.

Reply:
You're right. It's currently not in a single transaction but I will change it to be.